grakn-graql/src/main/java/ai/grakn/graql/internal/query/runner/TinkerComputeQueryRunner.java

```diff
@@ -0,0 +1,683 @@
+                }
+            }
+        }
+        return dataType;
+    }
+
+    private Set<Label> statisticsResourceLabels(StatisticsQuery<?> query) {
+        return calcStatisticsResourceTypes(query).stream()
+                .map(SchemaConcept::getLabel)
+                .collect(CommonUtil.toImmutableSet());
+    }
+
+    private ImmutableSet<Type> calcStatisticsResourceTypes(StatisticsQuery<?> query) {
+        if (query.attributeLabels().isEmpty()) {
+            throw GraqlQueryException.statisticsAttributeTypesNotSpecified();
+        }
+
+        return query.attributeLabels().stream()
+                .map((label) -> {
+                    Type type = this.tx.getSchemaConcept(label);
+                    if (type == null) throw GraqlQueryException.labelNotFound(label);
+                    if (!type.isAttributeType()) throw GraqlQueryException.mustBeAttributeType(type.getLabel());
+                    return type;
+                })
+                .flatMap(Type::subs)
+                .collect(CommonUtil.toImmutableSet());
+    }
+
+    private boolean selectedResourceTypesHaveInstance(ComputeQuery<?> query, Set<Label> statisticsResourceTypes) {
+        for (Label resourceType : statisticsResourceTypes) {
+            for (Label type : subLabels(query)) {
+                Boolean patternExist = tx.graql().infer(false).match(
+                        Graql.var("x").has(resourceType, Graql.var()),
+                        Graql.var("x").isa(Graql.label(type))
+                ).iterator().hasNext();
+                if (patternExist) return true;
+            }
+        }
+        return false;
+        //TODO: should use the following ask query when ask query is even lazier
+//        List<Pattern> checkResourceTypes = statisticsResourceTypes.stream()
+//                .map(type -> var("x").has(type, var())).collect(Collectors.toList());
+//        List<Pattern> checkSubtypes = subLabels.stream()
+//                .map(type -> var("x").isa(Graql.label(type))).collect(Collectors.toList());
+//
+//        return tx.get().graql().infer(false)
+//                .match(or(checkResourceTypes), or(checkSubtypes)).aggregate(ask()).execute();
+    }
+
+    private Set<Label> getCombinedSubTypes(StatisticsQuery<?> query) {
+        Set<Label> allSubTypes = getHasResourceRelationLabels(calcStatisticsResourceTypes(query));
+        allSubTypes.addAll(subLabels(query));
+        allSubTypes.addAll(query.attributeLabels());
+        return allSubTypes;
+    }
+
+    private Set<Label> getHasResourceRelationLabels(Set<Type> subTypes) {
+        return subTypes.stream()
+                .filter(Concept::isAttributeType)
+                .map(resourceType -> Schema.ImplicitType.HAS.getLabel(resourceType.getLabel()))
+                .collect(Collectors.toSet());
+    }// If the sub graph contains attributes, we may need to add implicit relations to the paths
+
+    private List<List<Concept>> getExtendedPaths(List<List<Concept>> allPaths) {
```

Comment:
the method is big but you seem to manage to split them into several high-level structure here. maybe actually split them into several methods?

Reply:
Yeah, I will try and split them more. `TinkerComputeQueryRunner` is biiiig...